_posts/2023-05-06-reverse-engineering-gba.markdown

```diff
@@ -0,0 +1,1062 @@
+* Destination: The register to set the value in. In this instruction, the value of `r0` will be set.
+* Source: The place to retrieve a value to store at the destination. In this case, the value is a constant value of 0; this is known as an __immediate__ value in assembly nomenclature.
+
+Combining all three parts, this instruction sets `r0` to a value of 0, discarding the previous value that was in `r0`.
+
+It is also possible for the source to be another register. The following instruction will copy the value of `r1` into `r0`.
+```
+mov r0,r1
+```
+
+> The extra 's' in the mnemonic `movs` is used for certain assembly languages to distinguish between different assignment operators that set flags or increment registers automatically, speeding up common operations like string assignment. There is no such distinction in THUMB assembly, so `movs` and `mov` are used interchangeably. This 's' also appears in other mnemonics.
+
+#### Arithmetic
+Let's look at a different kind of instruction. The instruction below is an addition operation.
+```
+adds r0,r1,#0x1
+```
+This instruction adds 1 to the value in `r1` and stores it in `r0`.
+
+`r0` is the destination register like before. Since addition requires two operands, there are two sources, `r1` and `#0x1`. It is possible to add either a register and an immediate value, or two registers. If the instruction uses an immediate value, the immediate value must always come last; this constraint has to do with how the instruction is implemented in the CPU's hardware.
+
+In the special case where you want to add an immediate value to a register value and store the new value back in the same register, you can use the following shorthand.
+```
+adds r0,#0x1
+```
+This will add 1 to the value of `r0` and store the sum in `r0`. Alternatively, you can say that this increments the value in `r0`.
+
+Other available math operators include subtraction, multiplication, negation, bitwise AND/OR/XOR/NOT, and logical/arithmetic (unsigned/signed) left/right bit shifts. Some of these operations are more rigid than addition, such as not supporting immediate values, though they all follow a similar structure. For a full list of supported instructions, refer to [ARM's developer documentation](https://developer.arm.com/documentation/ddi0210/c/Introduction/Instruction-set-summary/Thumb-instruction-summary?lang=en).
+
+>You may notice the lack of a division instruction. Division with arbitrary numbers is more complicated than the operations listed above, so it is implemented as a function rather than a single instruction. This means division is much slower than other math operations. Note that the right shift operator can divide numbers by powers of 2 to achieve certain divisions in a single instruction.
+
+#### Load/Store memory
+Registers can only store a handful of values, so memory is used to store the majority of a program's data.
```

Comment:
I would add a blurb here that memory here is RAM and it is where the remainder of the program values are stored.